app/src/main/kotlin/com/github/gotify/service/WebSocketService.kt

```diff
@@ -106,10 +104,47 @@ internal class WebSocketService : Service() {
             .onOpen { onOpen() }
             .onClose { onClose() }
             .onBadRequest { message -> onBadRequest(message) }
-            .onNetworkFailure { minutes -> onNetworkFailure(minutes) }
+            .onNetworkFailure { seconds -> onNetworkFailure(seconds) }
             .onMessage { message -> onMessage(message) }
             .onReconnected { notifyMissedNotifications() }
             .start()
+        return connection
+    }
+
+
+    private fun startPushService() {
+        UncaughtExceptionHandler.registerCurrentThread()
+        showForegroundNotification(getString(R.string.websocket_init))
+
+        //observe the network changes
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            (getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager)
```

Comment:
A previous version of gotify/android used network events to reconnect the connection. See https://github.com/gotify/android/pull/41 this change was ultimately reverted via https://github.com/gotify/android/issues/153 because the events weren't fired or fired after a big delay (probably when the phone left idle state).

I'm not confident enough to solely depend on the connection manager for reconnecting. Debugging problems that could be caused by this is really difficult, and the current implementation seems to run smoothly.

I think it'd be okay if we use the onAvailable callback to instantly reconnect (calling connection.start()), this way we wouldn't have to change much of the existing code, and only have another .start call to hopefully connect earlier than the reconnect timeout.

What do you think?

Reply:
Sure, we can just call `.start()` to perform a reconnect and do nothing when `.onLost()`. It's just want to save power to call `.close()` when `.onLost()`.

Reply:
system power manager may diminish the frequency of network data exchanges for background application, especially in power save mode. so I guess, that is why `the events weren't fired or fired after a big delay, but, here are two reasons why we can use `registerDefaultNetworkCallback` to perform a reconnection still.
- when `the events weren't fired or fired after a big delay, the creation of websocket will fail as well. so, trying to reconnect frequently will got the same effection as `events weren't fired`.
- `ConnectivityManager.NetworkCallback` is called by system, so It will be called finally whatever It is in power save mode or the network was really down.

Reply:
I don't think the connection should be closed when the network is lost. I'm not an expert there, but I'd expect this regularly happens when using mobile data. If the network is established fast enough, then we can use the still existing connection, because it may still be alive.

Additionally, the WebSocketConnection doesn't have to be recreated, because the class has functionality to reconnect.

> when "the events weren't fired or fired after a big delay", the creation of websocket will fail as well. so, trying to reconnect frequently will got the same effection as "events weren't fired"

I don't think it's the same effect. When using the timed reconnect, it is guaranteed to be called because the alarm manager is used. The experience with the ConnectionManager in #153 was, that it reconnects in some undefined amount of time.

Reply:
websocket based on TCP, a TCP connection was based on IP and port, and system will send FIN handshake package to close any TCP connections in the interface when It's down. and then system will assign a new random port for client(the next TCP connection's seq number will start with another index). So, whether the time slot is short or long, the system will always close the TCP connection. then websocket in application layer will recreate as well.

Reply:
but, websocket has the auto quick reconnection ability, ~~but, consider this situation: the client was down actually for a long time. in this way, server will wait in futile~~.all right, the network is down, so the disconnect(close signal for websocket) will sent failed as well, but, also consider this situation: the network will been down for a long time. Then the client will wait in futile too. so, we could stop call reconnect in our code(websocket will sent heart beat handshake package in a frequency by it's own). or just close the websocket to save more power.

Reply:
<img width="1728" alt="QQ20230502-193427@2x" src="https://user-images.githubusercontent.com/28947794/235655270-d20ac91a-5ee3-44ff-8e75-6662da9b9311.png">
<img width="1728" alt="QQ20230502-193509@2x" src="https://user-images.githubusercontent.com/28947794/235655295-423d2bd3-d810-4457-91a3-d5be61caa35c.png">

I dumped the interface for debug. I disconnected the network at 15.00 and 45.00. 
from the dump file, we can see that the server was still waiting for the client even though it had been down. and from the screenshot below, the client sent a FIN package to the server before creating a new connection.

<img width="1728" alt="QQ20230502-193533@2x" src="https://user-images.githubusercontent.com/28947794/235655661-fb68aaad-b49c-4298-883b-4d9a49e69c12.png">

Reply:
<img width="1728" alt="QQ20230502-200215@2x" src="https://user-images.githubusercontent.com/28947794/235660751-a2b0e270-5966-4f5d-bb4c-ecb02d4bb1f4.png">
I have taken another test. just let websocket reconnect when network resumed rather than close and establish a new websocket connection. This time, I just disconnected the wifi for 4 seconds. Although the TCP connection was recreated, but the websocket connection was still kept on. Maybe this solution will be better.

Reply:
The called `connection!!.scheduleReconnect(0)` for websocket connection will just reconnect rather than close and establish a new connection.

Reply:
Your adjusted solution is how I envisioned it. Tho, I find it difficult to understand how you got there.

I've made some changes, like moving the unregister to onDestroy, as stopService isn't actually called when the websocket service is stopped.

Reply:
Ok. Thanks.